plugins/migration/bin/rename.js

```diff
@@ -0,0 +1,286 @@
+
+      renamers.push(new VersionRenamer(database[version]));
+    }
+    return renamers;
+  }
+
+  /**
+   * Coerces the given string into a valid version (semver compliant or
+   * develop).
+   * @param {string} version  The version to coerce.
+   * @return  {string} The coerced version.
+   */
+  static coerce(version) {
+    return version === DEV_VERSION ?
+        version : semver.coerce(version).toString();
+  }
+
+  /**
+   * Compares the given versions. Compatible with Array.sort.
+   * @param {string} v1 The first version to compare.
+   * @param {string} v2 The second version to compare.
+   * @return {number} A number indicating the relationship between the versions.
+   */
+  static compare(v1, v2) {
+    if (v2 === DEV_VERSION) return -1;
+    if (v1 === DEV_VERSION) return 1;
+    return semver.compare(v1, v2);
+  }
+
+  /**
+   * Returns true if the first version is less than or equal to the second
+   * version.
+   * @param {string} v1 The version to compare.
+   * @param {string} v2 The version to compare against.
+   * @return {boolean} True if the first version is less than or equal to the
+   *     second one.
+   */
+  static lte(v1, v2) {
+    if (v2 === DEV_VERSION) return true;
+    if (v1 === DEV_VERSION) return false;
+    return semver.lte(v1, v2);
+  }
+
+  /**
+   * Returns true if the first version is greater than the second version.
+   * @param {string} v1 The version to compare.
+   * @param {string} v2 The version to compare against.
+   * @return {boolean} True if the first version is greater than the second one.
+   */
+  static gt(v1, v2) {
+    if (v2 === DEV_VERSION) return false;
+    if (v1 === DEV_VERSION) return true;
+    return semver.gtr(v1, `^${v2}`);
+  }
+
+  /**
+   * Applies the given renamings directly to a JavaScript string
+   * (presumably the contents of a developer's file).
+   * @param {string} str The string to apply the renamings in.
+   * @return {string} The file with renamings applied.
+   */
+  rename(str) {
+    return str.replace(dottedIdentifier, (match) => {
+      for (const versionRenamer of this.versionRenamers_) {
+        match = versionRenamer.rename(match);
+      }
+      return match;
+    });
+  }
+}
+
+/**
+ * An object which can apply a given set of renamings for a single
+ * Blockly version to the individual dotted-identifier paths.
+ */
+class VersionRenamer {
+  /**
+   * Constructor for the version renamer.
+   * @param {!Object} entry The database entry for a single version.
+   */
+  constructor(entry) {
+    /** 
+     * List of pre-compiled renamings.
+     * @private @const {
+     *     !Array<{old: string, new: ?string, get: ?string, set: ?string}>}
+     */ 
+    this.renamings_ = [];
+    // See the sample entry in renamings.json5 for explanation of the
+    // meaning of the different properties on database entries.
+    for (const module of entry) {
+      const oldModulePath = module.oldPath ?? module.oldName;
```

Comment:
Apparently null coalescing [doesn't exist in node twelve](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Nullish_coalescing_operator#browser_compatibility) :/ So I'm going to have to shuffle this around a bit.